GMP Precompile

jboetticher opened the pull request:
### Description

GMP Precompile documentation
Since it uses Wormhole, do you think it needs a disclaimer?

### Checklist

- [X] If this requires translations for the `moonbeam-docs-cn` repo, I have created a ticket for the translations in Jira
- [X] If images have been added, I have run the `compress-images.py` script to compress the images.
- [ ] If this page requires a disclaimer, I have added one

### After Translation Requirements

- [ ] Will need to create PR in `moonbeam-docs` repo to remove images
- [ ] Will need to create PR in `moonbeam-docs` repo to remove variables
- [ ] Will need to create PR in `moonbeam-mkdocs` repo to add redirects for Chinese site
- [ ] No additional PRs are required after the translations are done

eshaben commented on builders/interoperability/protocols/wormhole.md:
```diff
@@ -414,4 +414,10 @@ After a few seconds to a minute, cross-chain messages should be properly relayed
+
+Wormhole will provide liquidity to parachains through the Moonbeam Routed Liquidity program. This program allows one-click transfers of liquidity from Wormhole connected chains into parachain wallets by sending liquidity through Moonbeam networks.  
+
+MRL utilizes the [GMP Precompile](/builders/pallets-precompiles/precompiles/gmp), whose documentation explains how cross-chain messages should be constructed to properly use. Currently only available on Moonbase Alpha.  
```
```suggestion
MRL utilizes the [GMP Precompile](/builders/pallets-precompiles/precompiles/gmp){target=_blank}, whose documentation explains how cross-chain messages should be constructed to properly use the precompile. Currently, it is only available on Moonbase Alpha.  
```

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+
+![GMP Precompile Moonbeam Banner](/images/builders/pallets-precompiles/precompiles/gmp/gmp-banner.png)
+
+## Introduction {: #introduction } 
```
I think this introduction needs to be a bit more informative. Like what does GMP stand for? Maybe here is the best place to mention it's only available on Wormhole. 

Imagine if you had no prior knowledge and you stumbled upon this page - would you understand what it's about? I don't think you have to go into details on everything but including links saying you should be familiar with Wormhole or if you want to find more info out about XCM, check out the XCM docs. I see you mentioned briefly cross-chain protocols, that's a good spot to add a link to the protocols index page that we have

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+
+The most common instance that someone will have to interact with the precompile is in the case of a recovery, where a relayer doesn’t complete an MRL transaction. A user would have to find the VAA that comes with their origin chain transaction and then manually invoke the wormholeTransferERC20 function.
+
+## Building the Payload for Wormhole
```
header attributes are missing on some of these headers

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+
+Currently the GMP precompile only supports sending liquidity through Wormhole.  
+
+To send liquidity, on the origin chain, users must invoke the [`transferTokensWithPayload` method](https://book.wormhole.com/technical/evm/tokenLayer.html#contract-controlled-transfer){target=_blank} on the [WormholeTokenBridge smart contract](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/interfaces/ITokenBridge.sol){target=_blank}. It requires a bytes payloa ...
```
this guide focuses on moonbeam to other parachains correct? and not backwards? might be good to clearly specify the origin chain you're referring to is moonbeam?

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+- **AccountKey20** — an account id that is 20-bytes in length, including Ethereum-compatible account ids such as those on Moonbeam
+- **AccountId32** — an account id that is 32-bytes in length, standard in Polkadot and its parachains
+
+In the following MultiLocation templates, replace `INSERT_PARACHAIN_ID` with the parachain ID of the network you wish to send funds to and replace `ADDRESS_HERE` with the address of the account you want to send funds to.  
```
clearly state that this multilocation targets an account on another parachain, with moonbeam as the origin

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
````diff
@@ -0,0 +1,155 @@
+    }
+    ```
+
+It can be difficult to correctly SCALE encode the entire payload without the right tools. Fortunately, there are Polkadot JavaScript packages that can assist with this, such as `@polkadot/types`. The following script shows how to create a `Uint8Array` that can be used as a payload for the GMP precompile:  
````
i feel that this code snippet has a lot of code which is not explained

eshaben commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+                { "Parachain": INSERT_PARACHAIN_ID },
+                { 
+                    "AccountKey20": { 
+                        "network": "Any", 
```
network any? so this won't work for xcm v3?

jboetticher replied:
Both V2 and V3 are supported. Do you think it's worthwhile to convert the code into V3 @eshaben ?

jboetticher replied:
I updated to V3 and tested it out. It works well enough and is simpler, so I think I'll keep it at V3 in both the docs and the example I send to Wormhole. Thanks for pointing this out

eshaben replied:
ah sorry I missed the question til now, but sounds like a good plan 👍

jboetticher commented on builders/pallets-precompiles/precompiles/gmp.md:
```diff
@@ -0,0 +1,155 @@
+
+To send liquidity, on the origin chain, users must invoke the [`transferTokensWithPayload` method](https://book.wormhole.com/technical/evm/tokenLayer.html#contract-controlled-transfer){target=_blank} on the [WormholeTokenBridge smart contract](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/interfaces/ITokenBridge.sol){target=_blank}. It requires a bytes payloa ...
+
+You may be unfamiliar with both SCALE encoding and MultiLocations if you are not familiar with the Polkadot ecosystem. [SCALE encoding](https://docs.substrate.io/reference/scale-codec/){target=_blank} is a compact form of encoding that Polkadot uses. The [MultiLocation type](https://wiki.polkadot.network/docs/learn-xcvm){target=_blank} is used to define a relative point in Polkadot, such as a spe ...
```
We don't define multilocation in any of the other precompile interfaces, but to compensate I'm adding a link to a multilocation definition in the general XCM location. Do you think that would be enough for a reader? Most developers will likely just b-line it to the Wormhole section anyways

eshaben replied:
yeah the other XCM precompiles are a bit different since they're just sections embedded into pages which include the definition.

But now that you have pointed it out, I think this is something we can probably improve on. Maybe create an XCM-specific (or interoperability-specific) glossary page so it's easier to link to these definitions... 

Thanks for bringing it up! And yeah, your solution is good, thanks!

jboetticher commented:
Thanks for looking at this @eshaben , I've tried to add some additional clarity to the data types in the Wormhole relayer. Additionally, I replaced XCMv2 with XCMv3 since that's what we're going to be using moving forward. And last but not least, I've added additional comments within the code snippet to hopefully let readers know what's going on a bit better

eshaben reviewed (approved):
Beautifullllll! ✨
